src/VisualStudio/CSharp/Impl/Utilities/CSharpParseOptionsChangingService.cs

```diff
@@ -17,13 +20,25 @@ public CSharpParseOptionsChangingService()
         {
         }
 
-        public bool CanApplyChange(ParseOptions oldOptions, ParseOptions newOptions)
+        public bool CanApplyChange(ParseOptions oldOptions, ParseOptions newOptions, string? maxLangVersion)
         {
             var oldCSharpOptions = (CSharpParseOptions)oldOptions;
             var newCSharpOptions = (CSharpParseOptions)newOptions;
 
+            LanguageVersion parsedMaxLanguageVersion;
+
+            if (string.IsNullOrEmpty(maxLangVersion))
+            {
+                parsedMaxLanguageVersion = LanguageVersion.Latest;
+            }
+            else
+            {
+                Contract.ThrowIfFalse(LanguageVersionFacts.TryParse(maxLangVersion, out parsedMaxLanguageVersion));
+            }
+
             // Currently, only changes to the LanguageVersion of parse options are supported.
-            return oldCSharpOptions.WithLanguageVersion(newCSharpOptions.SpecifiedLanguageVersion) == newOptions;
+            return oldCSharpOptions.WithLanguageVersion(newCSharpOptions.SpecifiedLanguageVersion) == newOptions &&
+                newCSharpOptions.LanguageVersion <= parsedMaxLanguageVersion;
```

Comment:
Is it just easier moving the newCsharpOptions.LanguageVersion <= parsedMaxLanguageVersion in the "else" branch only? Then you don't need the sentinel value for the if case.

(The part that screwed me up was if maxLangVersion is unset, does this also include upgrading to preview versions, since Latest vs. Preview is a bit weird.)